.upstream-tests/test/std/utilities/expected/expected.expected/monadic/and_then.pass.cpp

```diff
@@ -0,0 +1,373 @@
+      assert(cuda::std::move(i).and_then(NOCRVal{}) == previous_error);
+      ASSERT_SAME_TYPE(decltype(cuda::std::move(i).and_then(CRVal{})), cuda::std::expected<int, TestError>);
+    }
+
+    //With & qualifier on F's operator()
+    {
+      const cuda::std::expected<int, TestError> i{0};
+      const RVCRefQual l{};
+      assert(i.and_then(cuda::std::move(l)) == 1);
+      const NORVCRefQual nl{};
+      assert(i.and_then(cuda::std::move(nl)) == expected_error);
+      ASSERT_SAME_TYPE(decltype(i.and_then(cuda::std::move(l))), cuda::std::expected<int, TestError>);
+    }
+
+    {
+      const cuda::std::expected<int, TestError> i{cuda::std::unexpect, 1337};
+      const RVCRefQual l{};
+      assert(i.and_then(cuda::std::move(l)) == previous_error);
+      const NORVCRefQual nl{};
+      assert(i.and_then(cuda::std::move(nl)) == previous_error);
+      ASSERT_SAME_TYPE(decltype(i.and_then(cuda::std::move(l))), cuda::std::expected<int, TestError>);
+    }
+  }
+}
+
+// For a generic lambda, nvrtc appears to not know what to do and claims it needs an annotation (when normal lambdas don't).
+// This is an expanded lambda from the original test.
+struct nvrtc_workaround {
+    template<typename T>
+    __host__ __device__
+    constexpr cuda::std::expected<int, TestError> operator()(T && t) {
+        return t.non_const();
+    }
+};
+
+// check that the lambda body is not instantiated during overload resolution
+__host__ __device__
+constexpr void test_sfinae() {
+  cuda::std::expected<NonConst, TestError> expect{};
+  auto l = nvrtc_workaround(); // [](auto&& x) { return x.non_const(); };
+  expect.and_then(l);
+  cuda::std::move(expect).and_then(l);
+}
+
+struct NoCopy {
+  NoCopy() = default;
+  __host__ __device__
+  constexpr NoCopy(const NoCopy&);
+  __host__ __device__
+  constexpr cuda::std::expected<int, TestError> operator()(const NoCopy&&) { return 1; }
+};
+
+// We need an indirection so the assert does not break the compilation
+template<class T>
+struct AlwaysFalse {
+  __host__ __device__
+  constexpr AlwaysFalse() { assert(false); }
+};
+
+struct NeverCalled {
+  template<class T>
+  __host__ __device__
+  constexpr cuda::std::expected<int, TestError> operator()(T) const { return AlwaysFalse<T>{}, cuda::std::expected<int, TestError>{42}; }
+};
+
+__host__ __device__
+constexpr bool test() {
+  test_val_types();
+  test_sfinae();
+
+  cuda::std::expected<int, TestError> expect{cuda::std::unexpect, 42};
+  const auto& cexpect = expect;
+
+  expect.and_then(NeverCalled{});
+  cuda::std::move(expect).and_then(NeverCalled{});
+  cexpect.and_then(NeverCalled{});
+  cuda::std::move(cexpect).and_then(NeverCalled{});
+
+  cuda::std::expected<NoCopy, TestError> nc{cuda::std::unexpect, 42};
+  const auto& cnc = nc;
+  cuda::std::move(cnc).and_then(NoCopy{});
+  cuda::std::move(nc).and_then(NoCopy{});
+
+  return true;
+}
+
+int main(int, char**) {
+  test();
+  static_assert(test(), "");
```

Comment:
I believe this is necessary. We want to ensure that we are indeed emplacing the function return directly into the right union member without going through `construct_at` 

If it would not work this would be a bug